File: OpenList-main/internal/db/image_favorites.go
package db

import (
	"github.com/OpenListTeam/OpenList/v4/internal/model"
	"github.com/pkg/errors"
)

// ImageFavoriteFolder operations

func ListImageFavoriteFoldersByUser(userId uint) ([]model.ImageFavoriteFolder, error) {
	var folders []model.ImageFavoriteFolder
	if err := db.Where("user_id = ?", userId).Order("`order` ASC, created_at DESC").Find(&folders).Error; err != nil {
		return nil, errors.Wrapf(err, "failed to list image favorite folders")
	}
	return folders, nil
}

func GetImageFavoriteFolderById(id uint, userId uint) (*model.ImageFavoriteFolder, error) {
	var folder model.ImageFavoriteFolder
	if err := db.Where("id = ? AND user_id = ?", id, userId).First(&folder).Error; err != nil {
		return nil, errors.Wrapf(err, "failed to get image favorite folder")
	}
	return &folder, nil
}

func CreateImageFavoriteFolder(folder *model.ImageFavoriteFolder) error {
	return errors.WithStack(db.Create(folder).Error)
}

func UpdateImageFavoriteFolder(folder *model.ImageFavoriteFolder) error {
	return errors.WithStack(db.Save(folder).Error)
}

func DeleteImageFavoriteFolderById(id uint, userId uint) error {
	// First delete all images in this folder
	if err := db.Where("folder_id = ? AND user_id = ?", id, userId).Delete(&model.ImageFavorite{}).Error; err != nil {
		return errors.Wrapf(err, "failed to delete images in folder")
	}

	// Then delete the folder
	if err := db.Where("id = ? AND user_id = ?", id, userId).Delete(&model.ImageFavoriteFolder{}).Error; err != nil {
		return errors.Wrapf(err, "failed to delete image favorite folder")
	}

	return nil
}

// ImageFavorite operations

func ListImageFavoritesByFolder(folderId uint, userId uint) ([]model.ImageFavorite, error) {
	var favorites []model.ImageFavorite
	if err := db.Where("folder_id = ? AND user_id = ?", folderId, userId).Order("created_at DESC").Find(&favorites).Error; err != nil {
		return nil, errors.Wrapf(err, "failed to list image favorites")
	}
	return favorites, nil
}

func ListAllImageFavoritesByUser(userId uint) ([]model.ImageFavorite, error) {
	var favorites []model.ImageFavorite
	if err := db.Where("user_id = ?", userId).Order("created_at DESC").Find(&favorites).Error; err != nil {
		return nil, errors.Wrapf(err, "failed to list all image favorites")
	}
	return favorites, nil
}

func GetImageFavoriteById(id uint, userId uint) (*model.ImageFavorite, error) {
	var favorite model.ImageFavorite
	if err := db.Where("id = ? AND user_id = ?", id, userId).First(&favorite).Error; err != nil {
		return nil, errors.Wrapf(err, "failed to get image favorite")
	}
	return &favorite, nil
}

func CheckImageFavoriteExists(userId uint, folderId uint, originalPath string) (bool, error) {
	var count int64
	if err := db.Model(&model.ImageFavorite{}).Where("user_id = ? AND folder_id = ? AND original_path = ?", userId, folderId, originalPath).Count(&count).Error; err != nil {
		return false, errors.Wrapf(err, "failed to check image favorite existence")
	}
	return count > 0, nil
}

func CreateImageFavorite(favorite *model.ImageFavorite) error {
	return errors.WithStack(db.Create(favorite).Error)
}

func UpdateImageFavorite(favorite *model.ImageFavorite) error {
	return errors.WithStack(db.Save(favorite).Error)
}

func DeleteImageFavoriteById(id uint, userId uint) error {
	if err := db.Where("id = ? AND user_id = ?", id, userId).Delete(&model.ImageFavorite{}).Error; err != nil {
		return errors.Wrapf(err, "failed to delete image favorite")
	}
	return nil
}

File: OpenList-Frontend-main/src/utils/image-favorites.ts
import { r } from "./request"
import type { ImageFavoriteFolder, ImageFavorite } from "~/types/image-favorite"

// Helper function to check response
const checkResp = (resp: any) => {
  if (resp.code !== 200) {
    throw new Error(resp.message || "Request failed")
  }
}

// Folder operations
export const listImageFavoriteFolders = async (): Promise<ImageFavoriteFolder[]> => {
  const resp = await r.get("/image_favorites/folder/list")
  checkResp(resp)
  return resp.data || []
}

export const getImageFavoriteFolder = async (id: number): Promise<ImageFavoriteFolder> => {
  const resp = await r.get("/image_favorites/folder/get", { params: { id } })
  checkResp(resp)
  return resp.data
}

export const createImageFavoriteFolder = async (data: {
  name: string
  description?: string
  order?: number
}): Promise<ImageFavoriteFolder> => {
  const resp = await r.post("/image_favorites/folder/create", data)
  checkResp(resp)
  return resp.data
}

export const updateImageFavoriteFolder = async (data: {
  id: number
  name?: string
  description?: string
  order?: number
}): Promise<ImageFavoriteFolder> => {
  const resp = await r.post("/image_favorites/folder/update", data)
  checkResp(resp)
  return resp.data
}

export const deleteImageFavoriteFolder = async (id: number): Promise<void> => {
  const resp = await r.post("/image_favorites/folder/delete", {}, { params: { id } })
  checkResp(resp)
}

// Image operations
export const listImageFavorites = async (folderId?: number): Promise<ImageFavorite[]> => {
  const resp = await r.get("/image_favorites/list", { 
    params: folderId ? { id: folderId } : {} 
  })
  checkResp(resp)
  return resp.data || []
}

export const createImageFavorite = async (data: {
  folder_id: number
  storage_id: number
  original_path: string
  file_name: string
  note?: string
  fingerprint?: string
}): Promise<ImageFavorite> => {
  const resp = await r.post("/image_favorites/create", data)
  checkResp(resp)
  return resp.data
}

export const updateImageFavorite = async (data: {
  id: number
  note: string
}): Promise<ImageFavorite> => {
  const resp = await r.post("/image_favorites/update", data)
  checkResp(resp)
  return resp.data
}

export const deleteImageFavorite = async (id: number): Promise<void> => {
  const resp = await r.post("/image_favorites/delete", {}, { params: { id } })
  checkResp(resp)
}

